Add kahan sum reduce helper class.

MrBurmark opened the pull request:
# Summary

This adds a KahanSum class to help do Kahan sums.
This also adds tests and generalizes the existing tests for reduce helper classes to accommodate slightly different interfaces.

The intention is to potentially use this or something similar for checksums in https://github.com/llnl/RAJAPerf/pull/608.

- This PR is a  feature
- It does the following:
  - Adds KahanSum class at the request of myself

MrBurmark committed:
Add kahan sum reduce helper class.

This adds a KahanSum class to help do Kahan sums.
This also adds tests and generalizes the existing tests for
reduce helper classes to accommodate slightly different interfaces.

artv3 commented:
Super nice! Can you add something in the RAJA docs about this new feature? And what do folks think about adding a note about this in our `RELEASE_NOTES` in this PR.

MrBurmark commented:
Actually we should probably not hide the KahanSum class in RAJA::detail if we want to use it in RAJAPerf.

MrBurmark committed:
Expose Reduce classes

Move them into RAJA samestpace from RAJA::detail namespace.
The intention is to use these in RAJAPerf

MrBurmark commented:
@artv3 I'll add an issue to add documentation, but I'd like to get this in so it can be used in https://github.com/llnl/RAJAPerf/pull/608.

MrBurmark committed:
fix style

artv3 commented on include/RAJA/policy/cuda/multi_reduce.hpp:
```diff
@@ -285,8 +285,8 @@ struct MultiReduceGridAtomicHostInit_TallyData
   {
-    ::RAJA::detail::HighAccuracyReduce<T, typename Combiner::operator_type>
-        reducer(m_identity);
+    ::RAJA::HighAccuracyReduce<T, typename Combiner::operator_type> reducer(
```
Why not just call it `KahanReduce`? What if we want to add other types of High Accuracy Reducers

MrBurmark replied:
I called it sum because it is always a sum and can't do generic operators.

artv3 commented on test/unit/algorithm/tests/test-algorithm-util-reduce.hpp:
```diff
@@ -199,6 +273,11 @@ using HipAccumulateReduceReducers =
+using HipKahanReduceReducers =
+  camp::list<
+              KahanSum<test_hip>
+            >;
```
I misread it, the policy name is KahanSum.

MrBurmark committed:
Merge branch 'develop' into feature/burmark1/kahan

MrBurmark committed:
Simplify implementation of reduce helper

rename clear to reset

MrBurmark committed:
Merge branch 'feature/burmark1/kahan' of github.com:LLNL/RAJA into feature/burmark1/kahan

MrBurmark committed:
fix style

MrBurmark committed:
Add const to reducer helper get methods
